packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,503 @@
+        }
+        // All other strings pass through to JSON
+        return passable;
+      }
+      case 'symbol': {
+        // At the smallcaps level, we prefix symbol names with `%`.
+        // By "symbol name", we mean according to `nameForPassableSymbol`
+        // which does some further escaping. See comment on that function.
+        assertPassableSymbol(passable);
+        const name = /** @type {string} */ (nameForPassableSymbol(passable));
+        return `%${name}`;
+      }
+      case 'undefined': {
+        return '#undefined';
+      }
+      case 'number': {
+        if (Number.isNaN(passable)) {
+          return '#NaN';
+        }
+        if (is(passable, -0)) {
+          return 0;
+        }
+        if (passable === Infinity) {
+          return '#Infinity';
+        }
+        if (passable === -Infinity) {
+          return '#-Infinity';
+        }
+        // All other numbers pass through to JSON
+        return passable;
+      }
+      case 'bigint': {
+        const str = String(passable);
+        return passable < 0n ? str : `+${str}`;
+      }
+      case 'copyRecord': {
+        // Currently copyRecord allows only string keys so this will
+        // work. If we allow sortable symbol keys, this will need to
+        // become more interesting.
+        const names = ownKeys(passable).sort();
+        return fromEntries(
+          names.map(name => [
+            encodeToSmallcapsRecur(name),
+            encodeToSmallcapsRecur(passable[name]),
+          ]),
+        );
+      }
+      case 'copyArray': {
+        return passable.map(encodeToSmallcapsRecur);
+      }
+      case 'tagged': {
+        return {
+          '#tag': encodeToSmallcapsRecur(getTag(passable)),
+          payload: encodeToSmallcapsRecur(passable.payload),
+        };
+      }
+      case 'remotable': {
+        const result = encodeRemotableToSmallcaps(
+          passable,
+          encodeToSmallcapsRecur,
+        );
+        if (typeof result === 'string' && result.startsWith('$')) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Remotable encoding must start with "$": ${result}`,
+        );
+      }
+      case 'promise': {
+        const result = encodePromiseToSmallcaps(
+          passable,
+          encodeToSmallcapsRecur,
+        );
+        if (typeof result === 'string' && result.startsWith('&')) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Promise encoding must start with "&": ${result}`,
+        );
+      }
+      case 'error': {
+        const result = encodeErrorToSmallcaps(passable, encodeToSmallcapsRecur);
+        if (
+          typeof result === 'object' &&
+          hasOwnPropertyOf(result, '#error') &&
+          typeof result['#error'] === 'string'
+        ) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Error encoding must have "#error" property: ${result}`,
+        );
```

Comment:
Consistency with the `encodeToSmallcaps` special case:
```suggestion
        if (typeof result === 'object' && hasOwnPropertyOf(result, '#error')) {
          const message = result['#error'];
          if (
            typeof message === 'string' &&
            // check that it decodes to a string
            (!startsSpecial(message) || message.startsWith('!'))
          ) {
            return result;
          }
          assert.fail(
            X`internal: Error encoding must string message: ${q(message)}`,
          );
        }
        assert.fail(
          X`internal: Error encoding must have "#error" property: ${q(result)}`,
        );
```
Also, when is `q` appropriate? Some calls to `X` use it but others don't... should it be limited to values that are expected to be strings?

Reply:
Nice symmetry. Done.

Using `q` or not expresses a security policy. Within a details (`X`) substitution, the contents are normally redacted, because the default assumption is that it may convey detailed data that is useful for debugging, but inappropriate to leak to callers. But when the information is relatively static an unlikely to convey a secret that callers should not see, use `q` to disclose it rather than redact it. 

A recent example illustrates this well: Redact amounts because they contain quantities. But brands and keywords should both be disclosed using `q`.